presto-main/src/main/java/io/prestosql/sql/planner/NullabilityAnalyzer.java

```diff
@@ -55,8 +55,17 @@ public static boolean mayReturnNullOnNonNullInput(Expression expression)
         @Override
         protected Void visitCast(Cast node, AtomicBoolean result)
         {
-            // try_cast and cast(JSON 'null' AS ...) can return null
-            result.set(true);
+            // Certain casts (e.g., cast(JSON 'null' AS ...)) can return null, but we know
+            // that any "type-only" coercion cannot produce null on non-null input, so
+            // take advantage of that fact to produce a more precise result.
+            //
+            // TODO: need a generic way to determine whether a CAST can produce null on non-null input.
+            // This should be part of the metadata associated with the CAST. (N.B. the rules in
+            // ISO/IEC 9075-2:2016, section 7.16.21 seems to imply that CAST cannot return NULL
+            // except for the CAST(NULL AS x) case -- we should fix this at some point)
```

Comment:
Is cast from JSON the only exception from the rule?

Reply:
I *think* so, but haven’t verified that’s the case

Reply:
In SQL:2016 there's first class support for JSON, and the way you turn a JSON value into a SQL null is by doing something like:

```
JSON_VALUE ('null', '$')
```

or 

```
JSON_VALUE ('null', '$') RETURNING BIGINT
```

if you wanted a BIGINT null, for instance.

There are no casts define for JSON in SQL:2016.